File: python/235.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def search(self, root, node, path):
        path.append(root)
        if root is node:
            return path
        elif node.val < root.val:
            return self.search(root.left, node, path)
        else:
            return self.search(root.right, node, path)

    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        ppath = self.search(root, p, [])
        qpath = self.search(root, q, [])
        return [i for i, j in zip(ppath, qpath) if i == j][-1]

File: cpp/235.cpp
#include<vector>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q) {
    vector<TreeNode*> path = {root};
    TreeNode* t = root;
    while(t != p){
      if(p->val < t->val)
        t = t->left;
      else
        t = t->right;
      path.push_back(t);
    }
    vector<TreeNode*> path2 = {root};
    t = root;
    while(t != q){
      if(q->val < t->val)
        t = t->left;
      else
        t = t->right;
      path2.push_back(t);
    }
    size_t i = 0;
    while(i < path.size() && i < path2.size()){
      if(path[i] == path2[i])
        t = path[i];
      else
        break;
      ++i;
    }
    return t;
  }
};
